src/main/scala/com/codacy/analysis/cli/Main.scala

```diff
@@ -38,40 +39,65 @@ class MainImpl extends CLIApp {
           "No credentials found.".asLeft[ProjectConfiguration]
         } { _.getRemoteConfiguration }
 
-        val resultsUploader: Either[String, ResultsUploader] = codacyClientOpt.fold {
-          "No credentials found.".asLeft[ResultsUploader]
-        } { codacyClient =>
-          analyse.commitUuid.fold {
-            "No commit option found.".asLeft[ResultsUploader]
-          } { commit =>
-            if (analyse.uploadValue) {
-              new ResultsUploader(commit, codacyClient, None).asRight[String]
-            } else {
-              "Upload option disabled.".asLeft[ResultsUploader]
-            }
-          }
-
-        }
-
-        new AnalyseExecutor(
+        val analysisResults = new AnalyseExecutor(
           analyse.tool,
           analyse.directory,
           formatter,
           analyser,
-          resultsUploader,
           fileCollector,
           remoteProjectConfiguration,
           analyse.parallel).run()
-    }
 
-    Await.result(res, Duration.Inf) match {
-      case Left(e) =>
-        logger.error(s"Failed analysis: $e")
-      case Right(_) =>
-        logger.info("Completed analysis successfully")
+        val uploadResult = uploadResults(codacyClientOpt)(analyse.uploadValue, analyse.commitUuid, analysisResults)
+        Await.result(uploadResult, Duration.Inf) match {
+          case Left(err) if analyse.uploadValue =>
```

Comment:
We can move this "if" to before the match, so we don't even run the "uploadResults" function if "analyse.uploadValue" is false.